### PR TITLE
Auto/launcher ground segmentation

### PR DESCRIPTION
**Pull request by satoshi-ota**
## PR Type

- .auto porting
- support for ground segmentation
- related PR -> https://github.com/tier4/AutowareArchitectureProposal_aip_launcher/pull/2

## Review Procedure

- please check topic connection by using `rqt_graph`

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Code follows [coding guidelines][coding-guidelines]
- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code follows [coding guidelines][coding-guidelines]
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets
- [ ] Write [release notes][release-notes]

## CI Checks

- **Build and test for PR**: Required to pass before the merge.
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[coding-guidelines]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/1194394777/T4
[release-notes]: https://tier4.atlassian.net/wiki/spaces/AIP/pages/563774416
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    with open(path, "r") as f:
+        p = yaml.safe_load(f)
+    return p
+
+
+def create_additional_pipeline(vehicle_info, lidar_name):
+    cropbox_component = ComposableNode(
+        package="pointcloud_preprocessor",
+        plugin="pointcloud_preprocessor::CropBoxFilterComponent",
+        name=lidar_name + "_crop_box_filter",
```
Comment (taichiH) [inline perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:57]: Both are okay but following style is a common.
```
f"{lidar_name}_crop_box_filter"
```

If you change coding style, please apply same change to other lines.

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+                "local_slope_max_angle_deg": 30.0,
+                "split_points_distance_tolerance": 0.2,
+                "split_height_distance": 0.2,
+                "use_virtual_ground_point": False,
+            }
+        ],
+        extra_arguments=[{"use_intra_process_comms": LaunchConfiguration("use_intra_process")}],
+    )
+
+    load_laserscan_to_occupancy_grid_map = IncludeLaunchDescription(
```
Comment (taichiH) [inline perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:194]: please rename `load_laserscan_to_occupancy_grid_map` to `laserscan_to_occupancy_grid_map_loader`

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        target_container=container,
+        condition=IfCondition(
+            LaunchConfiguration(
+                "use_additional_pipeline", default=pipeline_param["use_additional_pipeline"]
+            )
+        ),
+    )
+
+    reference_component_loader = LoadComposableNodes(
+        composable_node_descriptions=[ground_component],
```
Comment (taichiH) [inline perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:256]: What is `ground_component`?

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py (hunk 1 above)
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/c8731fb573aa47f554296bba0fae0074adc7c249

File: perception_launch/config/object_segmentation/ground_segmentation/aip_xx1_perception.param.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+/**:
+  ros__parameters:
+    use_additional_pipeline: False
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception_launch/config/object_segmentation/ground_segmentation/aip_xx1_perception.param.yaml]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/fefa905458cf78edf1df07ecfebeefe7bed40d2e

File: perception_launch/config/object_segmentation/ground_segmentation/aip_x2_perception.param.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+/**:
+  ros__parameters:
+    use_additional_pipeline: True
+    additional_lidar: ["front_upper", "front_lower"]
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception_launch/config/object_segmentation/ground_segmentation/aip_x2_perception.param.yaml]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/fefa905458cf78edf1df07ecfebeefe7bed40d2e

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py (hunk 2 above)
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py]: fix inhttps://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/92d24c1db2184978dd4fb8c144bf4ab388d0cc68

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py (hunk 3 above)
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py]: It is a ground filter node, but the name is not readable.
fix components name in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/3f2a0dbd8c0bcb84a7121da15438c816b4b67970

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+    ground_concat_topics = ["concatenated/no_ground/pointcloud"]
+    for lidar_name in pipeline_param["additional_lidar"]:
+        ground_concat_topics.extend([lidar_name + "/no_ground/pointcloud"])
+
+    ground_concat_component = ComposableNode(
+        package="pointcloud_preprocessor",
+        plugin="pointcloud_preprocessor::PointCloudConcatenateDataSynchronizerComponent",
+        name="concatenate_data_front",
+        remappings=[("output", "no_ground/oneshot/pointcloud")],
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:164]: The components to be launched depend on whether additional pipelines are used or not.

use additional pipeline:

- **ground_filter_component_use_additional_pipeline**, and
- **concat_data_component** to be launched

**not** use additional pipeline:

- **ground_filter_component** to be launched

So, **ground_filter_component_use_additional_pipeline** and **ground_filter_component** are **NOT** launched at the same time, and the output topic names (`no_ground/oneshot/pointcloud`) are same to connect the next laserscan_to_occupancy_grid_map node.

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py (hunk 3 above)
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/732f49b305177898d8256dd8699b871997f04418

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+                "max_y": 50.0,
+                "min_z": vehicle_info["min_height_offset"],
+                "max_z": vehicle_info["max_height_offset"],
+                "negative": False,
+            }
+        ],
+        extra_arguments=[{"use_intra_process_comms": LaunchConfiguration("use_intra_process")}],
+    )
+
+    customized_ground_component = ComposableNode(
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/732f49b305177898d8256dd8699b871997f04418

File: perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:164]: fix in https://github.com/tier4/AutowareArchitectureProposal_launcher/pull/117/commits/732f49b305177898d8256dd8699b871997f04418

[the remark this answers is not in the document]
Comment (taichiH) [reply perception_launch/launch/object_segmentation/ground_segmentation/ground_segmentation.launch.py:164]: Thanks!

**Review by taichiH — APPROVED** [review]
LGTM

**Comment by yukkysaito** [comment]
@taichiH @satoshi-ota Did you get it to work?

**Comment by satoshi-ota** [comment]
I have checked this branch by using rosbag data.
It looks good to me.

rviz_screenshot_2021_11_18-14_08_18

**Comment by yukkysaito** [comment]
Thank you !

**Pull request merged**